m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,831 @@
+        CountValue countValue = new CountValue();
+        countValue.setI64Value(value);
+
+        MetricValue metricValue = new MetricValue();
+        metricValue.setCount(countValue);
+
+        Metric metric = newMetric(name, tags, metricValue);
+
+        queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+    }
+
+    private Metric newMetric(String name, Map<String, String> tags, MetricValue metricValue) {
+        Metric metric = new Metric(name);
+        metric.setTags(toMetricTagSet(tags));
+        metric.setTimestamp(System.currentTimeMillis() * Duration.NANOS_PER_MILLI);
+        metric.setMetricValue(metricValue);
+
+        return metric;
+    }
+
+    private void queueSizedMetric(SizedMetric sizedMetric) {
+        try {
+            metricQueue.put(sizedMetric);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            LOG.warn(String.format("Interrupted queueing metric: %s", sizedMetric.getMetric().getName()));
+        }
+    }
+
+    private class Processor implements Runnable {
+        List<Metric> metrics = new ArrayList<>(freeBytes / 10);
+        int metricsSize = 0;
+
+        @Override
+        public void run() {
+            try {
+                while (!executor.isShutdown()) {
+                    // This `take` call will block until there is an item on the queue to take.
+                    // When this reporter is closed, shutdownNow will be called on the executor,
+                    // which will interrupt this thread and proceed to the `InterruptedException`
+                    // catch block.
+                    SizedMetric sizedMetric = metricQueue.take();
+
+                    flushMetricIteration(sizedMetric);
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } finally {
+                flushRemainingMetrics();
+
+                // Always arrive at phaser to prevent deadlock
+                phaser.arrive();
+            }
+        }
+
+        private void flushMetricIteration(SizedMetric sizedMetric) {
+            Metric metric = sizedMetric.getMetric();
+
+            if (sizedMetric == SizedMetric.FLUSH) {
+                flush(metrics);
+
+                return;
+            }
+
+            int size = sizedMetric.getSize();
+
+            if (metricsSize + size > freeBytes) {
+                flush(metrics);
+                metricsSize = 0;
+            }
+
+            metrics.add(metric);
+
+            metricsSize += size;
+        }
+
+        private void flushRemainingMetrics() {
+            while (!metricQueue.isEmpty()) {
+                flushMetricIteration(metricQueue.remove());
+            }
+
+            flush(metrics);
+        }
+    }
+
+    /**
+     * Builder pattern to construct an {@link M3Reporter}.
+     */
+    public static class Builder {
```

Comment:
nit: could you move the Builder above the private methods of the class

Reply:
Checkstyle actually specifies that inner classes go after variables and methods 🤔

Reply:
Personally I prefer ordering by access. The builder will/can be used by external callers. The internal non-public methods cannot. Not a big deal if you will publish javadocs, you plan to do that before releasing yeah?

Reply:
I actually have another branch that already automatically publishes javadocs here:
https://uber-java.github.io/tally/docs/

It will be in another PR. I'll look at that other hosting solution if applicable after publishing to Maven Central.

Reply:
Sounds good